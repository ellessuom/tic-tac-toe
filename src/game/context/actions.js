export const PLAYERS_ID = {
  PLAYER_ONE: 'p1',
  PLAYER_TWO: 'p2',
};

export const INIT = {
  canPlay: true, // allows to add new items to the matrix
  isTie: false, // no winners - all tiles filled
  currentPlayer: PLAYERS_ID.PLAYER_ONE, // user id, either 'p1' or 'p2'
  usedTiles: []
};

export default (state, dispatch) => {
  return ({
    startGame: () => {
      dispatch({
        ...INIT,
        [PLAYERS_ID.PLAYER_ONE]: {
          ...state[PLAYERS_ID.PLAYER_ONE],
          selectedTiles: []
        },
        [PLAYERS_ID.PLAYER_TWO]: {
          ...state[PLAYERS_ID.PLAYER_TWO],
          selectedTiles: []
        }
      });
    },
    play: (tileIdx) => {
      dispatch({
        usedTiles: [...state.usedTiles, tileIdx].sort(),
        [state.currentPlayer]: {
          ...state[state.currentPlayer],
          selectedTiles: [...state[state.currentPlayer].selectedTiles, tileIdx].sort(),
        },
        currentPlayer: state.currentPlayer === PLAYERS_ID.PLAYER_TWO? PLAYERS_ID.PLAYER_ONE : PLAYERS_ID.PLAYER_TWO
      });
    },
  });
};
