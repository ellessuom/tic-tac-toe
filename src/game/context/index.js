import Provider, { useData } from './context';

export { Provider, useData };
