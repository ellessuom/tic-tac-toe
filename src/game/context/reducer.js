import {
  INIT,
  PLAYERS_ID
} from './actions';
import Cookies from "universal-cookie";

export const initialState = {
  ...INIT,
  [PLAYERS_ID.PLAYER_ONE]: {
    wins: 0,
    selectedTiles: []
  },
  [PLAYERS_ID.PLAYER_TWO]: {
    wins: 0,
    selectedTiles: []
  },
};

export default (state, payload) => {
  const cookies = new Cookies();
  const newState = {...state, ...payload };
  cookies.set('ttt', newState);
  return newState;
};
