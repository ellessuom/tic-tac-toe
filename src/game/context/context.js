import React from 'react';
import PropTypes from 'prop-types';
import reducer, { initialState } from './reducer';
import availableActions from './actions';
import Cookies from "universal-cookie";

const Context = React.createContext();

const DataProvider = ({ children }) => {
  const cookies = new Cookies();
  const [state, dispatch] = React.useReducer(reducer, cookies.get('ttt'), initialState);
  const actions = availableActions(state, dispatch);

  const value = React.useMemo(() => [{ ...state }, { ...actions }], [
    state, actions,
  ]);

  React.useEffect(() => {
    if (state.usedTiles) {
      actions.verify();
    }

  }, [state.usedTiles]);

  return <Context.Provider value={value}>{children}</Context.Provider>;
};

export function useData() {
  const context = React.useContext(Context);
  if (!context) {
    throw new Error(
      'You must use Data Provider in order to consume this context.',
    );
  }

  return context;
}

export default DataProvider;

DataProvider.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired,
};
