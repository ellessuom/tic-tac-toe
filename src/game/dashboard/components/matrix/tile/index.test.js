import React from 'react';
import { mount } from 'enzyme';
import Tile from './index';
import { useData } from "../../../../context";

const TILE_IDX = 3;
const fakeUseData = [
  {
    canPlay: true,
    usedTiles: [],
    winnerPattern: '',
    currentPlayer: 'p1',
    p1: {
      selectedTiles: []
    },
    p2: {
      selectedTiles: []
    }
  }
];

jest.mock('../../../../context');

describe('<Tile />', () => {
  let wrapper;
  const fakeActions = { play: jest.fn()};
  const init = (payload) => {
    useData.mockImplementation(() => payload);
    return mount(<Tile i={TILE_IDX}/>);
  };

  beforeAll(() => {
    wrapper = init(fakeUseData);
  });

  it('should render the default tile', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have default \'available\' class', () => {
    expect(wrapper.find('div').hasClass('available')).toEqual(true);
  });

  describe('onClick - available tile', () => {
    const generateFakeData = (playerId) => ({
      canPlay: true,
      usedTiles: [],
      winnerPattern: '',
      currentPlayer: playerId,
    });


    describe('Player One', () => {
      beforeAll(() => {
        wrapper = init([generateFakeData('p1'), fakeActions]);
        wrapper.find('div').simulate('click');
      });

      it('should have \'used\' class', () => {
        expect(wrapper.find('div').hasClass('used')).toEqual(true);
      });

      it(`should contain 'X' mark`, () => {
        expect(wrapper.find('img').props().alt).toEqual('X');
      });
    });

    describe('Player Two', () => {
      beforeAll(() => {
        wrapper = init([generateFakeData('p2'), fakeActions]);
        wrapper.find('div').simulate('click');
      });

      it('should have \'used\' class', () => {
        expect(wrapper.find('div').hasClass('used')).toEqual(true);
      });

      it(`should contain 'O' mark`, () => {
        expect(wrapper.find('img').props().alt).toEqual('O');
      });
    });
  });

  describe('onClick - used tile', () => {
    const generateFakeData = (playerId) => ({
      canPlay: true,
      usedTiles: [TILE_IDX],
      winnerPattern: '',
      currentPlayer: playerId,
      p1: {
        selectedTiles: [TILE_IDX]
      }
    });

    beforeAll(() => {
      wrapper = init([generateFakeData('p1'), fakeActions]);
      wrapper.find('div').simulate('click');
    });

    it('should have \'used\' class when clicked once', () => {
      expect(wrapper.find('div').hasClass('used')).toEqual(true);
    });

    it(`should not have changed when clicked again`, () => {
      wrapper.find('div').simulate('click');
      expect(wrapper).toMatchSnapshot();
    });

    describe('Player Two', () => {
      beforeAll(() => {
        wrapper = init([generateFakeData('p2'), fakeActions]);
        wrapper.find('div').simulate('click');
      });

      it('should have \'used\' class', () => {
        expect(wrapper.find('div').hasClass('used')).toEqual(true);
      });

      it(`should not have changed when clicked again`, () => {
        wrapper.find('div').simulate('click');
        expect(wrapper).toMatchSnapshot();
      });
    });
  });
});
