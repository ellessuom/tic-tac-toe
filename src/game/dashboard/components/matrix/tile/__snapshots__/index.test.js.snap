// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tile /> onClick - used tile Player Two should not have changed when clicked again 1`] = `
<Tile
  i={3}
>
  <div
    className="tile used"
    onClick={[Function]}
  >
    <img
      alt="X"
      src="mark_x.svg"
      width={50}
    />
  </div>
</Tile>
`;

exports[`<Tile /> onClick - used tile should not have changed when clicked again 1`] = `
<Tile
  i={3}
>
  <div
    className="tile used"
    onClick={[Function]}
  >
    <img
      alt="X"
      src="mark_x.svg"
      width={50}
    />
  </div>
</Tile>
`;

exports[`<Tile /> should render the default tile 1`] = `
<Tile
  i={3}
>
  <div
    className="tile available p1"
    onClick={[Function]}
  />
</Tile>
`;
